content/blog/2023-12-11-array-ir/index.md

```diff
@@ -0,0 +1,109 @@
+
+The idea behind this algorithm is that we “regret” choosing to execute vertex $v$ on a particular protocol $P$ at time $t$ when some descendant $v’$ of $v$ is on protocol $P$, but has a dependency $d$ which is on a different protocol $P_d$ and was ready at time $t$ — that is, we could have chosen to execute $d$ at time $t$. By choosing $v$ at time $t$, we are forced to pause execution on protocol $P$ and switch to $P_d$ to evaluate $d$ before returning to $P$ to execute $v’$. To minimize this “regret”, when we have a choice of which protocol to choose next, we try to choose one which will resolve the greatest number of dependencies on other protocols.
+
+Reordering performs a topological sort of the nodes; the main portion of the algorithm is the heuristic to determine which is the next node to process (`FetchNext`). 
+
+If we are currently on a certain protocol (lines 1-3), then we will complete any nodes that are ready on that protocol. If there are no nodes left, we choose the next protocol by determining which protocol will “free up” the most dependencies. 
+Lines 5-8 find all the reachable nodes for each node in `ready`. Lines 9-12 compute protocols of all nodes that will be resolved if we process a certain protocol. Line 13 then selects the protocol which will resolve the greatest number of dependencies.
+
+This is still a work in progress. We have an idea for a better heuristic which may even produce provably optimum solutions.
+
+## Evaluation
+To rigorously evaluate our re-ordering heuristic, we made a mock-up of this problem at this [Github Repo](https://github.com/willwng/colored-tree-grouping). 
+
+This Python code generates a random DAG and assigns a random protocol to each node. The code finds the optimal solution by brute-forcing every topological ordering of the nodes and finding the minimum number of protocol changes required; the limitation of these tests is that they are not very scalable and we could only realistically test on small DAG’s. We then compare this result to our heuristic and find the following “correctness” distribution:
+<p align="center">
+    <img src="error_histogram.png" alt="error_histogram" width=450 />
+</p>
+
+Here, the error is defined by the number of swaps computed by our algorithm compared to the optimal solution. `p` is defined to be the probability of a DAG edge existing. 
+
+From these results, we see that our heuristic gives an optimal solution nearly 90% (at least for smaller DAG’s). We also estimate that the smallest approximation factor that describes our algorithm is `ɑ=2`, where our algorithm gives twice as many swaps as the optimal solution (number is obtained by a small adversarial example).
```

Comment:
"nearly 90%": add "…of the time"? Or make it "for nearly 90% of the randomly generated graphs"?